### PR TITLE
feat: add JaxOptOptimizer for end-to-end differentiable optimization

### PR DESCRIPTION
**Pull request by ericchansen**
## Summary

Adds **`JaxOptOptimizer`** — an end-to-end analytical optimizer that compiles the whole objective (force field → MM engine → loss → gradient) into a single `jax.jit`-traced graph and hands it to `jaxopt.LBFGS` / `LBFGSB` / `GradientDescent`. This removes the Python↔JAX round-trip from every iteration.

**This PR delivers Phase 3 of both #152 and #176** — it does not yet close them. See "What's still missing" below.

### New modules
- `q2mm/optimizers/spec.py` (222L) — `ObjectiveSpec` / `MoleculeSpec` frozen JAX-compatible snapshots, pre-bucketed by evaluator category.
- `q2mm/optimizers/jaxloss.py` (287L) — `JaxLoss._loss_fn`, a JIT-traced scalar loss over all molecules, with cached `jax.value_and_grad`.
- `q2mm/optimizers/jaxopt_opt.py` (227L) — `JaxOptOptimizer` with `method={"lbfgs","lbfgsb","gradient_descent"}`, `maxiter`, `tol`, `verbose`.
- `q2mm/models/hessian.py` — new `_jax_frequency_param_jacobian` implementing the closed-form `dλ/dp = vᵀ(dH/dp)v` eigenvalue derivative, fully JIT-compatible, plus `symbols_to_masses_3n` helper for jittable mass lookup.

### Dispatch + integration
- `ObjectiveFunction.to_jax_spec()` — bridge from Python objective to `ObjectiveSpec`.
- `OptimizationLoop` now accepts `jaxopt:lbfgs`, `jaxopt:lbfgsb`, `jaxopt:gradient_descent` as the gradient phase of grad-simp cycling.
- `q2mm-benchmark` picks up `jaxopt:lbfgs` and `jaxopt:lbfgsb` in the default matrix.
- `ensure_jaxopt()` lazy-import helper; `jaxopt` added to the `jax` and `gpu` extras (Windows excluded, matching JAX-MD).

### Benchmarks (CH₃F, committed under `benchmarks/ch3f/`)

| Form | Device | Optimizer | RMSD (cm⁻¹) | Notes |
|---|---|---|---:|---|
| harmonic | CPU | `jaxopt:lbfgsb` (A) | **528.3** | best harmonic |
| harmonic | GPU | SciPy L-BFGS-B (A) | 528.7 | reference |
| harmonic | CPU | `jaxopt:lbfgs` (A) | 531.9 | — |
| mm3 | GPU | `jaxopt:lbfgs` (A) | 578.7 | — |
| mm3 | GPU | SciPy L-BFGS-B (A) | 579.0 | reference |

Solution quality matches SciPy within < 1 cm⁻¹. CH₃F is too small for XLA fusion to beat Fortran L-BFGS — the speedup win is on larger/batched problems and is deferred to the vmap phase.

### Tests
- `test/test_jax_loss.py` (532L) — `ObjectiveSpec`, `JaxLoss` behavior, frequency/Hessian/eigenmatrix parity vs `ObjectiveFunction`, multi-molecule, mixed objectives, frequency-sensitivity parity vs NumPy.
- `test/test_jaxopt_optimizer.py` (338L) — validation, convergence, bounds-active LBFGSB, frequency convergence.
- `test/integration/test_jaxopt.py` (371L) — cycling dispatch, CH₃F, water Hessian, SN2 TS validation.
- Core suite: **606 passed, 261 skipped (backend-gated), 0 failures.**

### Docs
- Workflow D in `optimization-guide.md` (end-to-end differentiable).
- Step 6d in `tutorial.md`.
- Architecture diagram updated (`architecture.md`).
- Benchmark evidence in `small-molecules.md` (split into main + composed-workflows tables, matrix size 82).
- `AGENTS.md` now documents the canonical docs workflow: `uv sync --extra docs` + `uv run properdocs serve -f properdocs.yml`.

## What's still missing

This PR is **Phase 3** of both tracking issues. It does **not** close them. Remaining work, tracked in subsequent PRs:

- **Geometry refs via implicit differentiation** (#152 Phase 2) — `ObjectiveSpec.has_geometry_refs` is wired to `False`; `to_jax_spec()` raises when only geometry refs are present. Requires differentiation through a JAX-native inner energy minimizer.
- **`vmap` molecule batching** (#176 Phase 2) — `JaxLoss._loss_fn` still Python-unrolls over `mol_data`. The named GPU-speedup goal of #176 lands in the vmap PR.
- **TS curvature inversion (QFUERZA) inside JIT** — currently runs outside the compiled path; blocks rh-enamide / SN2 TS parameterization on the analytical pipeline.
- **MM3 backend full JIT** — OOP bends, stretch-bend cross terms, high-order torsions still have NumPy fallbacks.
- **Multi-start / basin-hopping as JAX primitives** — currently orchestrated from Python.
- **Parameter constraints as JAX projections** — equivalences, links, type-specific bounds still handled Python-side.

## Testing

- `python -m ruff check q2mm/ test/ scripts/` ✅
- `python -m ruff format --check q2mm test scripts examples` ✅
- `python -m pytest test/ -x -q -m "not (openmm or tinker or jax or jax_md or psi4)"` → 606 passed, 0 failed.
- Backend-gated tests (marked `jax`, `openmm`, etc.) run in CI.

## Refs

Refs #152 (Phase 3), #176 (Phase 3).

### CODE REVIEW

File: q2mm/diagnostics/benchmark.py [hunk 1]
```diff
@@ -765,6 +767,32 @@ def run_combo(
...
+        jaxopt_kwargs: dict[str, Any] = {
+            "method": jaxopt_method,
+            "maxiter": maxiter,
+            "verbose": False,
+        }
+        for key in ("tol", "stepsize", "linesearch"):
+            if key in optimizer_kwargs:
+                jaxopt_kwargs[key] = optimizer_kwargs[key]
+
+        opt = JaxOptOptimizer(**jaxopt_kwargs)
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/diagnostics/benchmark.py:794]: Fixed in d810c28. Dropped `stepsize`/`linesearch` from the forwarded kwargs — they weren't accepted by `JaxOptOptimizer.__init__` and no current CLI path sets them. Only `tol` is forwarded now.

File: q2mm/diagnostics/benchmark.py [hunk 2]
```diff
@@ -608,7 +608,9 @@ def run_combo(
             ``'Powell'`` dispatch to :class:`ScipyOptimizer`; ``'cycling'``
             dispatches to :class:`OptimizationLoop`; ``'optax:adam'`` (or
             ``'optax:adagrad'``, ``'optax:sgd'``, ``'optax:adamw'``)
-            dispatches to :class:`OptaxOptimizer`.
+            dispatches to :class:`OptaxOptimizer`; ``'jaxopt:lbfgs'`` (or
+            ``'jaxopt:lbfgsb'``, ``'jaxopt:gd'``) dispatches to
+            :class:`JaxOptOptimizer`.
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/diagnostics/benchmark.py:613]: Fixed in d810c28. Docstring now says `jaxopt:gradient_descent` to match `_METHOD_REGISTRY`.

File: docs/tutorial.md [hunk 3]
```diff
@@ -769,6 +769,46 @@ outperform L-BFGS-B. On CH₃F with MM3, Adam achieves **56.3 cm⁻¹ RMSD** —
...
+a single JIT-traced function. This enables true second-order methods like
+L-BFGS and L-BFGS-B with exact gradients.
+
+???+ example "JaxOpt L-BFGS-B optimization"
+
+    ```python
+    from q2mm.optimizers.jaxopt_opt import JaxOptOptimizer
+
+    optimizer = JaxOptOptimizer(
+        method="lbfgsb",   # or "lbfgs", "gd"
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply docs/tutorial.md]: Fixed in d810c28. Example comment now reads `method='gradient_descent'` to match the registry.

File: q2mm/optimizers/jaxopt_opt.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+            final_score=final_score,
+            n_iterations=n_iter,
+            n_evaluations=n_iter,
+            initial_params=x0,
+            final_params=final_params,
+            history=[initial_score, final_score],
+            method=method_str,
+            jac_mode="jit",
+            eps=None,
+        )
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/optimizers/jaxopt_opt.py:203]: Fixed in d810c28. `jac_mode="analytical"` now — gradients really are analytical (autodiff through the residual via `value_and_grad`), and `_resolve_gradients` / `benchmark_stem` label them correctly as `_auto`. Also renamed the 6 committed `*_fd.json` artifacts to `*_auto.json` with `jac_mode='analytical'` in their metadata so stems and stored labels agree.

File: q2mm/optimizers/spec.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+            hess_vals.append(ref.value)
+            hess_wts.append(ref.weight)
+        elif ref.kind == "eig_diagonal":
+            ediag_idx.append(ref.data_idx)
+            ediag_vals.append(ref.value)
+            ediag_wts.append(ref.weight)
+        elif ref.kind == "eig_offdiagonal":
+            eoff_idx.append(ref.data_idx)
+            eoff_vals.append(ref.value)
+            eoff_wts.append(ref.weight)
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/optimizers/spec.py:206]: Fixed in d810c28. `_build_molecule_spec` now packs `atom_indices=(row, col)` as `row * 3N + col` for `eig_offdiagonal` refs (mirrors the `hessian_element` branch), matching the decode on `jaxloss.py`.

File: q2mm/optimizers/spec.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        hess_weights: ``(n_hess,)`` weights for Hessian residuals.
+        eig_diag_indices: ``(n_ediag,)`` indices into the eigenvalue
+            vector for diagonal eigenmatrix references.
+        eig_diag_refs: ``(n_ediag,)`` reference eigenvalue values.
+        eig_diag_weights: ``(n_ediag,)`` weights for diagonal residuals.
+        eig_offdiag_indices: ``(n_eoff, 2)`` row/column pairs for
+            off-diagonal eigenmatrix references.
+        eig_offdiag_refs: ``(n_eoff,)`` reference off-diagonal values.
+        eig_offdiag_weights: ``(n_eoff,)`` weights for off-diagonal
+            residuals.
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/optimizers/spec.py]: Fixed in d810c28. `MoleculeSpec` docstring now describes `hess_indices` and `eig_offdiag_indices` as packed 1D `(row * 3N + col)` arrays, matching the implementation and the JaxLoss decode.

File: q2mm/optimizers/jaxopt_opt.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+                initial_score,
+                self.maxiter,
+            )
+
+        # Run the solver (LBFGSB needs bounds passed to run())
+        if self.method == "lbfgsb":
+            lower = jnp.array(spec.lower_bounds, dtype=jnp.float64)
+            upper = jnp.array(spec.upper_bounds, dtype=jnp.float64)
+            result = solver.run(params, bounds=(lower, upper))
+        else:
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/optimizers/jaxopt_opt.py:162]: Fixed in d810c28. Added a `jax.default_backend()` check at the top of the LBFGSB branch that raises a clear `RuntimeError` suggesting either `method='lbfgs'` on GPU or forcing CPU with `jax.config.update('jax_platform_name', 'cpu')`.

File: q2mm/optimizers/jaxloss.py [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+
+                    # Frequency contribution
+                    if mol_spec.has_frequency:
+                        dh_dp_kcal = jax.jacrev(hess_fn, argnums=1)(flat_coords, params)
+                        dh_dp_au = dh_dp_kcal * hess_au_scale
+
+                        freqs, d_freq_dp = _jax_frequency_param_jacobian(hess_au, dh_dp_au, entry["masses_3n"])
+                        calc_freqs = freqs[entry["freq_indices"]]
+                        residuals = entry["freq_weights"] * (entry["freq_refs"] - calc_freqs)
+                        total = total + jnp.sum(residuals**2)
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/optimizers/jaxloss.py:202]: Fixed in d810c28. Removed the `jax.jacrev(hess_fn, argnums=1)` call — `d_freq_dp` really was unused, and `value_and_grad` computes the param gradient via reverse-mode AD through the residual. Added a small `_jax_frequencies_from_hessian` helper that just eigh's the mass-weighted Hessian and returns frequencies. This should be a significant speedup.

File: q2mm/optimizers/jaxloss.py [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+
+                    # Hessian element contribution
+                    if mol_spec.has_hessian:
+                        if not mol_spec.has_frequency:
+                            dh_dp_kcal = jax.jacrev(hess_fn, argnums=1)(flat_coords, params)
+                            dh_dp_au = dh_dp_kcal * hess_au_scale
+
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/optimizers/jaxloss.py]: Fixed in d810c28. The Hessian-element branch no longer computes `dH/dp` either — same reasoning as the frequency branch above. Gradient flows entirely through `value_and_grad` of the residual.
